File: src/auto_tagger/integrations/fallback.py
"""Folder-structure fallback lookup helpers."""

from __future__ import annotations

import re
from pathlib import Path

from auto_tagger.core import iter_audio_files, read_metadata
from auto_tagger.core.metadata import TrackMetadata
from auto_tagger.core.parse_filename import parse_album_folder_name, parse_track_filename
from auto_tagger.exceptions import AutoTaggerError
from auto_tagger.integrations.candidates import (
    AlbumCandidate,
    LookupRequest,
    LookupSource,
    TrackCandidate,
)

# Common patterns in folder names that aren't part of the album name
_DATE_PREFIX_RE = re.compile(r"^(\d{4})[-.](?:0[1-9]|1[0-2])\s*")  # "2003-06" or "2005.08"
_YEAR_PREFIX_RE = re.compile(r"^(\d{4})[.-]\s*")  # "2017-" or "2018." alone
_YEAR_FROM_PREFIX_RE = re.compile(r"^(\d{4})[-.]")  # capture year from date prefix (e.g. "2003" from "2003-04《挚爱》")
_BOOKMARKS_RE = re.compile(r"[《》「」【】\[\]]")  # Chinese/Western bookmarks
_EXTRA_SUFFIX_RE = re.compile(r"\s*\([^)]*\)\s*$")  # trailing "(FLAC分轨)" etc.
_FORMAT_SUFFIX_RE = re.compile(r"\[?(flac|mp3|wav|aac|ogg|m4a|wma|ape|flac\s*分轨|wav\s*分轨)\]?\s*$", re.IGNORECASE)  # trailing [flac] etc.
_CD_SUBFOLDER_RE = re.compile(r"(?:[Cc][Dd]|[Dd][Ii][Ss][CcKk]|ディスク)\s*\d+\s*$")  # CD1, Disc 1, Disk1, ディスク1


def extract_year_from_name(name: str) -> str | None:
    """Extract a 4-digit year from a folder name.

    Tries, in order:
      1. Leading date prefix: "2003-04《挚爱》" → "2003"
      2. Inside bookmark brackets: "Artist-《2011-重译》" → "2011"
      3. Inside parentheses: "Artist (2011) Album" → "2011"

    "2005.08 Album" → "2005"
    "2017- Album" → "2017"
    "Album Name" → None

    Returns the year string (e.g. "2003") or None.
    """
    # 1. Leading date prefix
    match = _YEAR_FROM_PREFIX_RE.match(name)
    if match:
        return match.group(1)
    # 2. Inside Chinese bookmarks: 《2011-重译》
    m = re.search(r'[《（（\[]\s*(\d{4})\s*[-.]', name)
    if m:
        return m.group(1)
    # 3. Parenthesized year anywhere: (2011) or [2011]
    m = re.search(r'[\[(（]\s*(\d{4})\s*[\])）]', name)
    if m:
        return m.group(1)
    return None


def clean_folder_name(name: str) -> str:
    """Clean a folder name for use as a lookup hint.

    Strips common metadata prefixes/suffixes that aren't part of the
    actual album or artist name:
      - Date prefixes: "2003-04《挚爱》" → "挚爱"
      - Bookmarks: "《挚爱》" → "挚爱"
      - Extra suffixes: "Hello (Bonus)" → "Hello"

    For folder names containing paired bookmarks (e.g. "Artist-《2011-Album》[FLAC]"),
    the content inside the bookmarks is extracted first, then cleaned.
    This ensures the album hint is just the album name without artist/year prefixes.

    Returns the cleaned name, or the original if nothing to strip.
    """
    # First try extracting content from inside Chinese bookmarks:
    # "Artist-《2011-Album》[FLAC]" → extract "2011-Album" → clean → "Album"
    bracketed = re.search(r'《([^》]+)》', name)
    if bracketed:
        inner = bracketed.group(1)
        cleaned = _DATE_PREFIX_RE.sub("", inner)
        cleaned = _YEAR_PREFIX_RE.sub("", cleaned)
        cleaned = cleaned.strip()
        if cleaned:
            return cleaned

    # Fallback: standard cleanup on full name
    cleaned = _DATE_PREFIX_RE.sub("", name)
    cleaned = _YEAR_PREFIX_RE.sub("", cleaned)
    cleaned = _BOOKMARKS_RE.sub("", cleaned)
    # Strip format suffix first so parenthetical suffix can be matched
    cleaned = _FORMAT_SUFFIX_RE.sub("", cleaned)
    cleaned = _EXTRA_SUFFIX_RE.sub("", cleaned)
    cleaned = cleaned.strip()
    return cleaned or name


def parse_album_path(path: Path) -> LookupRequest:
    """Parse artist, album, and year hints from an Artist/Album path.

    Folder names are cleaned of date prefixes, bookmarks, and extra
    suffixes before being used as lookup hints. The year is extracted
    from the album folder's leading date prefix if present.
    """
    album_path = path.parent if path.is_file() else path
    album_hint = clean_folder_name(album_path.name) if album_path.name else None

    # Detect CD subfolder pattern (e.g. "Artist - Album CD1" or
    # "Artist.Album CD2").  When inside a CD subfolder the immediate
    # parent is the album bundle folder ("Album (2CD)"), and the
    # grandparent is the true artist folder.
    if album_path.name and _CD_SUBFOLDER_RE.search(album_path.name):
        artist_hint = (
            clean_folder_name(album_path.parent.parent.name)
            if album_path.parent.parent and album_path.parent.parent.name
            else None
        )
    else:
        artist_hint = clean_folder_name(album_path.parent.name) if album_path.parent.name else None

    year_hint = extract_year_from_name(album_path.name) if album_path.name else None
    return LookupRequest(
        path=path,
        artist_hint=artist_hint,
        album_hint=album_hint,
        year_hint=year_hint,
    )


def parse_album_with_tags(path: Path) -> LookupRequest:
    """Build a LookupRequest using both folder names and existing file tags.

    Existing tag values take priority over folder names — a human (or
    previous tool) put those tags there intentionally. Folder names are
    only used as fallback hints when tags are missing.

    Exception: if tag artist clearly differs from the folder artist name,
    the folder name wins (tags may have been written incorrectly by a
    previous auto-tag run).
    """
    folder_request = parse_album_path(path)
    tag_hints = _read_album_tags_from_first_file(path)

    # Use tag hints, but verify artist matches folder structure
    tag_artist = tag_hints.get("artist")
    folder_artist = folder_request.artist_hint

    if tag_artist and folder_artist and tag_artist.casefold() != folder_artist.casefold():
        # Tag artist doesn't match folder — folder structure is more trustworthy.
        # Also distrust the album tag since it may have been written by a wrong run.
        artist_hint = folder_artist
        album_hint = folder_request.album_hint
    else:
        artist_hint = tag_artist or folder_artist
        # Prefer cleaned folder album hint over tag — folder names are more
        # recently trustworthy and may have been cleaned of suffixes.
        album_hint = folder_request.album_hint or tag_hints.get("album")
    tracks = folder_request.tracks or _track_hints_from_path(path)

    return LookupRequest(
        path=path,
        artist_hint=artist_hint,
        album_hint=album_hint,
        year_hint=folder_request.year_hint or tag_hints.get("year"),
        tracks=tracks,
    )


def _read_album_tags_from_first_file(path: Path) -> dict[str, str | None]:
    """Read album-level tags from the first audio file in the directory."""
    try:
        audio_paths = iter_audio_files(path)
    except AutoTaggerError:
        return {}
    for audio_path in audio_paths:
        metadata = _read_metadata_or_none(audio_path)
        if metadata is not None and (metadata.album or metadata.artist):
            return {
                "artist": metadata.artist or metadata.album_artist,
                "album": metadata.album,
                "year": metadata.year,
            }
    return {}


def candidate_from_folder(request: LookupRequest) -> AlbumCandidate:
    """Build a low-confidence fallback candidate from folder and file hints.

    Enriches ``LookupRequest`` hints with structured folder-name parsing.
    Compilation detection is handled upstream by the workflow.
    """
    tracks = request.tracks or _track_hints_from_path(request.path)

    # Enrich hints using structured folder-name parsing
    album_path = request.path.parent if request.path.is_file() else request.path
    parsed_folder = parse_album_folder_name(album_path.name) if album_path.name else None

    # Use parsed values as fallback when request hints are missing
    artist = request.artist_hint or (parsed_folder.artist if parsed_folder else None)
    album = request.album_hint or (parsed_folder.album if parsed_folder else None)
    year = request.year_hint or (parsed_folder.year if parsed_folder else None)
    album_artist = artist

    return AlbumCandidate(
        artist=artist,
        artists=[artist] if artist else [],
        album=album,
        album_artist=album_artist,
        album_artists=[album_artist] if album_artist else [],
        year=year,
        tracks=tracks,
        source=LookupSource.FOLDER,
    )


def _tag_or_parsed(meta_value, parsed_value, fallback=None):
    """Return *meta_value* if truthy, else *parsed_value*, else *fallback*."""
    return meta_value if meta_value else (parsed_value if parsed_value else fallback)


def _track_hints_from_path(path: Path) -> list[TrackCandidate]:
    """Build track candidates from audio file metadata and filename parsing.

    Priority: embedded tags > filename parsing > raw file stem / index.
    """
    try:
        audio_paths = iter_audio_files(path)
    except AutoTaggerError:
        return []

    tracks: list[TrackCandidate] = []
    for index, audio_path in enumerate(audio_paths, start=1):
        m = _read_metadata_or_none(audio_path)
        p = parse_track_filename(audio_path.stem)

        tracks.append(
            TrackCandidate(
                title=_tag_or_parsed(m.title if m else None, p.title, audio_path.stem),
                artist=_tag_or_parsed(m.artist if m else None, p.artist),
                artists=(m.artists if m and m.artists else p.artists),
                track_number=_tag_or_parsed(m.track_number if m else None, p.track_number, index),
                track_total=len(audio_paths),
                disc_number=_tag_or_parsed(m.disc_number if m else None, p.disc_number),
                disc_total=m.disc_total if m else None,
                musicbrainz_trackid=m.musicbrainz_trackid if m else None,
            )
        )
    return tracks


def _read_metadata_or_none(path: Path) -> TrackMetadata | None:
    try:
        return read_metadata(path)
    except AutoTaggerError:
        return None

File: frontend/electron/handlers/fallback.ts
/**
 * Folder-structure fallback lookup helpers.
 * Ported from Python auto_tagger.integrations.fallback.
 *
 * Parses a file path into artist/album/year hints using folder names,
 * existing audio tags, and filename patterns.
 */

import { readdirSync, statSync } from "node:fs";
import { join, dirname, basename, extname } from "node:path";
import { readTrackMetadata } from "./tracks";
import {
  type AlbumCandidate,
  type LookupRequest,
  type TrackCandidate,
  makeAlbumCandidate,
  makeTrackCandidate,
} from "./candidates";

// ── Regex patterns ──────────────────────────────────────────────────

const COMPILATION_FOLDER_SET = new Set([
  "compilations",
  "compilation",
  "various artists",
  "various",
  "va",
  "soundtracks",
  "soundtrack",
  "ost",
  "samplers",
  "sampler",
  "christmas",
]);

const AUDIO_EXTENSIONS = new Set([".mp3", ".flac", ".m4a", ".mp4", ".wav", ".ogg", ".opus", ".ape"]);

function isAudioFilePath(inputPath: string): boolean {
  try {
    return statSync(inputPath).isFile();
  } catch {
    const name = basename(inputPath);
    if (/[《》「」【】]/.test(name)) return false;
    return AUDIO_EXTENSIONS.has(extname(name).toLowerCase());
  }
}

/**
 * Check if a folder name indicates a compilation/sampler rather than a single artist.
 */
export function isCompilationFolder(name: string | null | undefined): boolean {
  if (!name) return false;
  const normalized = name.trim().toLowerCase().replace(/[ _]+/g, " ");
  return COMPILATION_FOLDER_SET.has(normalized);
}

/** Album artist used when the folder indicates a compilation. */
const VARIOUS_ARTISTS = "Various Artists";

const DATE_PREFIX_RE = /^(\d{4})[-.](?:0[1-9]|1[0-2])(?!\d)(?:[-.](?:0[1-9]|[12]\d|3[01]))?\s*/; // "2003-04", "2007-09-28", or "2005.08"
const YEAR_PREFIX_RE = /^(\d{4})\s*[.-]\s*/; // "2017-" or "2018."
const YEAR_FROM_PREFIX_RE = /^(\d{4})[-.]/; // capture year from "2003-"
const STANDALONE_YEAR_RE = /(?:^|[^\d])((?:19|20)\d{2})(?!\d)/;
const BOOKMARKS_RE = /[《》「」【】\[\]]/g;
const EXTRA_SUFFIX_RE = /\s*\([^)]*\)\s*$/;
const FORMAT_SUFFIX_RE = /\[?(flac|mp3|wav|aac|ogg|m4a|wma|ape|flac\s*分轨|wav\s*分轨)\]?\s*$/i;
const CD_SUBFOLDER_RE = /(?:[Cc][Dd]|[Dd][Ii][Ss][CcKk]|ディスク)\s*\d+\s*$/;
const EDITION_KEYWORDS_RE = /\s*(?:香港首版|台湾首版|引进版|日本版|欧版|美版|内地版|中国大陆版|大陆版|德国版|澳洲版|新加坡版|马来西亚版|韩版)\s*/gi;
const DISC_COUNT_RE = /\s*\d+\s*(?:CD|Disc|ディスク)\s*$/i;

// ── Year extraction ─────────────────────────────────────────────────

/**
 * Extract a 4-digit year from a folder name.
 * Tries leading date prefix, inside bookmarks, or parenthesized.
 */
export function extractYearFromName(name: string): string | null {
  // 1. Leading date prefix
  const m1 = YEAR_FROM_PREFIX_RE.exec(name);
  if (m1) return m1[1];

  // 2. Inside Chinese bookmarks: 《2011-重译》
  const m2 = /[《（（\[]\s*(\d{4})\s*[-.]/.exec(name);
  if (m2) return m2[1];

  // 3. Parenthesized year: (2011) or [2011]
  const m3 = /[\[(（]\s*(\d{4})\s*[\])）]/.exec(name);
  if (m3) return m3[1];

  // 4. Standalone year in names like "黄绮珊《时光》2018 .wav".
  const m4 = STANDALONE_YEAR_RE.exec(name);
  if (m4) return m4[1];

  return null;
}

// ── Folder name cleaning ────────────────────────────────────────────

/**
 * Clean a folder name for use as a lookup hint.
 * Strips date prefixes, bookmarks, edition annotations, format suffixes.
 */
export function cleanFolderName(name: string): string {
  // First try extracting content from inside Chinese bookmarks
  const bracketed = name.match(/《([^》]+)》/);
  if (bracketed) {
    let inner = bracketed[1];
    inner = inner.replace(DATE_PREFIX_RE, "");
    inner = inner.replace(YEAR_PREFIX_RE, "");
    inner = inner.replace(EDITION_KEYWORDS_RE, "");
    inner = inner.replace(DISC_COUNT_RE, "");
    inner = inner.trim();
    if (inner) return inner;
  }

  // Fallback: standard cleanup on full name
  let cleaned = name.replace(DATE_PREFIX_RE, "");
  cleaned = cleaned.replace(YEAR_PREFIX_RE, "");
  cleaned = cleaned.replace(BOOKMARKS_RE, "");
  cleaned = cleaned.replace(EDITION_KEYWORDS_RE, "");
  cleaned = cleaned.replace(FORMAT_SUFFIX_RE, "");
  cleaned = cleaned.replace(EXTRA_SUFFIX_RE, "");
  cleaned = cleaned.replace(DISC_COUNT_RE, "");
  cleaned = cleaned.trim();
  return cleaned || name;
}

/**
 * Clean an album folder name for use as the album hint.
 * Strips the leading "Year - " prefix.
 */
export function cleanAlbumFolderName(
  name: string,
  artistHint?: string | null,
): string {
  let cleaned = cleanFolderName(name);

  const artist = artistHint ? cleanFolderName(artistHint).trim() : "";
  if (artist) {
    const afterArtist = stripRepeatedArtistFromAlbumFolder(cleaned, artist);
    if (afterArtist) return afterArtist;
  }

  // Also strip leading "Year - " prefix that's not just a date prefix
  const yearDash = /^(\d{4})\s*[-—]\s*/.exec(cleaned);
  if (yearDash) {
    cleaned = cleaned.slice(yearDash[0].length);
  }
  return cleaned;
}

function normalizeArtistPrefix(name: string): string {
  return name.toLowerCase().replace(/[\s._:：\-—–]+/g, "");
}

function cleanAlbumRemainder(name: string): string {
  let cleaned = name.trim().replace(/^[\s._:：\-—–]+/, "");
  cleaned = cleanFolderName(cleaned);
  cleaned = cleaned.replace(/^[\s._:：\-—–]+/, "").trim();
  return cleaned;
}

function stripRepeatedArtistFromAlbumFolder(
  albumFolder: string,
  artist: string,
): string | null {
  for (const prefix of repeatedArtistPrefixes(albumFolder, artist)) {
    const afterPrefix = albumFolder.slice(prefix.length);
    if (
      looksLikeArtistAlbumSeparator(afterPrefix, true) ||
      (prefix !== artist && /^\s+/.test(afterPrefix))
    ) {
      const cleaned = cleanAlbumRemainder(afterPrefix);
      if (cleaned) return stripEnglishArtistFromRemainder(cleaned, artist) ?? cleaned;
    }
  }

  const artistIndex = albumFolder.indexOf(artist);
  if (artistIndex <= 0) return null;

  const beforeArtist = albumFolder.slice(0, artistIndex);
  const afterArtist = albumFolder.slice(artistIndex + artist.length);
  if (
    looksLikeCategoryArtistSeparator(beforeArtist) &&
    looksLikeArtistAlbumSeparator(afterArtist, false)
  ) {
    const cleaned = cleanAlbumRemainder(afterArtist);
    if (cleaned) return stripEnglishArtistFromRemainder(cleaned, artist) ?? cleaned;
  }

  return null;
}

function repeatedArtistPrefixes(
  albumFolder: string,
  artist: string,
): string[] {
  const prefixes = new Set<string>();
  if (albumFolder.startsWith(artist)) prefixes.add(artist);

  const normalizedArtist = normalizeArtistPrefix(artist);
  if (!normalizedArtist) return [...prefixes];

  const firstSpace = albumFolder.search(/\s/);
  if (firstSpace > 0) {
    const firstToken = albumFolder.slice(0, firstSpace);
    if (normalizeArtistPrefix(firstToken).startsWith(normalizedArtist)) {
      prefixes.add(firstToken);
    }
  }

  return [...prefixes].sort((a, b) => b.length - a.length);
}

function looksLikeCategoryArtistSeparator(text: string): boolean {
  return /(?:\s{2,}|[._:：\-—–])\s*$/.test(text);
}

function looksLikeArtistAlbumSeparator(
  text: string,
  allowYearAfterSingleSpace: boolean,
): boolean {
  if (/^\s{2,}/.test(text)) return true;
  if (/^\s*[-—–]\s+/.test(text)) return true;
  if (/^[._:：\-—–]+/.test(text)) return true;
  return allowYearAfterSingleSpace && /^\s+(?:19|20)\d{2}\b/.test(text);
}

/**
 * When the artist is pure CJK (no Latin) and the album remainder starts
 * with an English name that is the artist's English name, strip it.
 * e.g. "George Lam Ultimate Sound Vol. II" → "Ultimate Sound Vol. II"
 * when artist is "林子祥" (George Lam's Chinese name).
 */
function stripEnglishArtistFromRemainder(
  cleaned: string,
  artist: string,
): string | null {
  if (artist.match(/[\u4e00-\u9fff]/) && !artist.match(/[a-zA-Z]/)) {
    const latinLead = cleaned.match(
      /^([A-Z][a-zA-Z]+(?:\s+[A-Z][a-zA-Z]+)?)\s+(.+)/,
    );
    if (latinLead) return latinLead[2];
  }
  return null;
}

// ── Path parsing ────────────────────────────────────────────────────

/**
 * Parse artist, album, and year hints from an Artist/Album path.
 */
export function parseAlbumPath(filePath: string): LookupRequest {
  const isFile = isAudioFilePath(filePath);
  const albumPath = isFile ? dirname(filePath) : filePath;
  const albumName = basename(albumPath);
  const parentName = basename(dirname(albumPath));

  let artistHint: string | null = null;
  let albumHint: string | null = null;
  let yearHint: string | null = extractYearFromName(albumName);

  // Detect CD subfolder pattern
  if (CD_SUBFOLDER_RE.test(albumName)) {
    const grandparent = basename(dirname(dirname(albumPath)));
    artistHint = grandparent ? cleanFolderName(grandparent) : null;
    const parent = basename(dirname(albumPath));
    albumHint = parent ? cleanAlbumFolderName(parent, artistHint) : null;
    yearHint = parent ? extractYearFromName(parent) : yearHint;
  } else {
    artistHint = parentName ? cleanFolderName(parentName) : null;
    albumHint = cleanAlbumFolderName(albumName, artistHint) || null;
  }

  return {
    path: filePath,
    artistHint,
    albumHint,
    yearHint,
    musicbrainzAlbumId: null,
    musicbrainzArtistId: null,
    discogsReleaseId: null,
    discogsArtistId: null,
    tracks: [],
  };
}

/**
 * Build a LookupRequest using both folder names and existing file tags.
 * Tag values take priority over folder names.
 */
export async function parseAlbumWithTags(filePath: string): Promise<LookupRequest> {
  const folderRequest = parseAlbumPath(filePath);
  const scanned = await scanAlbumFilesWithTags(filePath);

  const tagArtist = scanned.artist;
  const folderArtist = folderRequest.artistHint;

  let artistHint: string | null;
  let albumHint: string | null;

  if (isCompilationFolder(folderArtist)) {
    // Albums under "Compilations" folders are various-artist compilations
    artistHint = VARIOUS_ARTISTS;
    albumHint = folderRequest.albumHint;
  } else if (tagArtist && folderArtist && tagArtist.toLowerCase() !== folderArtist.toLowerCase()) {
    // Tag artist doesn't match folder — folder structure is more trustworthy
    artistHint = folderArtist;
    albumHint = folderRequest.albumHint;
  } else {
    artistHint = tagArtist || folderArtist;
    albumHint = folderRequest.albumHint || scanned.album;
  }

  return {
    path: filePath,
    artistHint,
    albumHint,
    yearHint: folderRequest.yearHint || scanned.year,
    musicbrainzAlbumId: scanned.musicbrainzAlbumId,
    musicbrainzArtistId: scanned.musicbrainzArtistId,
    discogsReleaseId: scanned.discogsReleaseId,
    discogsArtistId: scanned.discogsArtistId,
    tracks: scanned.tracks,
  };
}

interface AlbumFileScan {
  artist: string | null;
  album: string | null;
  year: string | null;
  musicbrainzAlbumId: string | null;
  musicbrainzArtistId: string | null;
  discogsReleaseId: string | null;
  discogsArtistId: string | null;
  tracks: TrackCandidate[];
}

async function scanAlbumFilesWithTags(filePath: string): Promise<AlbumFileScan> {
  const isFile = isAudioFilePath(filePath);
  const dirPath = isFile ? dirname(filePath) : filePath;
  const result = {
    artist: null as string | null,
    album: null as string | null,
    year: null as string | null,
    musicbrainzAlbumId: null as string | null,
    musicbrainzArtistId: null as string | null,
    discogsReleaseId: null as string | null,
    discogsArtistId: null as string | null,
    tracks: [] as TrackCandidate[],
  } satisfies AlbumFileScan;

  try {
    const entries = readdirSync(dirPath).sort();
    let index = 0;
    for (const entry of entries) {
      const fullPath = join(dirPath, entry);
      try {
        if (!statSync(fullPath).isFile()) continue;
      } catch {
        continue;
      }
      const ext = extname(entry).toLowerCase();
      if (!AUDIO_EXTENSIONS.has(ext)) continue;
      index++;
      const filenameHint = trackHintFromFilename(entry);

      try {
        const meta = await readTrackMetadata(fullPath);
        result.artist ??= meta.artist || meta.albumArtist;
        result.album ??= meta.album;
        result.year ??= meta.year;
        result.musicbrainzAlbumId ??= meta.musicbrainzAlbumId;
        result.musicbrainzArtistId ??= meta.musicbrainzArtistId;
        result.discogsReleaseId ??= meta.discogsReleaseId;
        result.discogsArtistId ??= meta.discogsArtistId;
        result.tracks.push(
          makeTrackCandidate({
            title: meta.title || filenameHint.title,
            artist: meta.artist || filenameHint.artist,
            artists: meta.artists.length > 0 ? meta.artists : filenameHint.artists,
            trackNumber: filenameHint.trackNumber ?? meta.trackNumber ?? index,
            trackTotal: null, // filled after all tracks are collected
            discNumber: meta.discNumber,
            musicbrainzTrackId: meta.musicbrainzTrackId,
            length: meta.duration,
            genre: meta.genre,
          }),
        );
      } catch {
        // Read failed — use filename as title hint
        result.tracks.push(
          makeTrackCandidate({
            title: filenameHint.title,
            artist: filenameHint.artist,
            artists: filenameHint.artists,
            trackNumber: filenameHint.trackNumber ?? index,
          }),
        );
      }
    }
  } catch {
    // directory can't be read
  }

  const total = result.tracks.length;
  for (const track of result.tracks) {
    track.trackTotal = total;
  }

  return result;
}

// ── Track hints ─────────────────────────────────────────────────────

/**
 * Build track candidates from audio file metadata.
 */
export async function trackHintsFromPath(filePath: string): Promise<TrackCandidate[]> {
  return (await scanAlbumFilesWithTags(filePath)).tracks;
}

function trackHintFromFilename(
  filename: string,
): { title: string; artist: string | null; artists: string[]; trackNumber: number | null } {
  const ext = extname(filename);
  const stem = filename.slice(0, filename.length - ext.length).trim();
  const match = /^(\d{1,3})\s*[.\-_\s]+\s*(.+)$/.exec(stem);
  const trackNumber = match ? Number(match[1]) : null;
  const withoutTrackNumber = match ? match[2].trim() || stem : stem;
  const artistTitle = splitFilenameArtistTitle(withoutTrackNumber);
  if (artistTitle) {
    return {
      title: cleanFilenameTitle(artistTitle.title),
      artist: artistTitle.artist,
      artists: [artistTitle.artist],
      trackNumber,
    };
  }
  return {
    title: cleanFilenameTitle(withoutTrackNumber),
    artist: null,
    artists: [],
    trackNumber,
  };
}

function splitFilenameArtistTitle(stem: string): { artist: string; title: string } | null {
  const separator = /\s[-–—]\s/.exec(stem);
  if (!separator || separator.index === 0) return null;
  const artist = stem.slice(0, separator.index).trim();
  const title = stem.slice(separator.index + separator[0].length).trim();
  if (!artist || !title) return null;
  return { artist, title };
}

function cleanFilenameTitle(title: string): string {
  const original = title.trim();
  let cleaned = original.replace(/\s*\([^)]*(?:bit|hz|khz|wav|flac|mp3|ape)[^)]*\)\s*$/i, "").trim();
  if (cleaned !== original) {
    cleaned = cleaned.replace(/\s*\([^)]*\)\s*$/u, "").trim();
  }
  return cleaned || title.trim();
}

// ── Folder fallback candidate ───────────────────────────────────────

/**
 * Build a low-confidence fallback candidate from folder and file hints.
 */
export function candidateFromFolder(request: LookupRequest): AlbumCandidate {
  const artist = request.artistHint;
  const album = request.albumHint;
  const year = request.yearHint;

  const isCompilation = isCompilationFolder(artist);
  const albumArtist = isCompilation ? VARIOUS_ARTISTS : artist;
  const albumArtists = isCompilation
    ? [VARIOUS_ARTISTS]
    : artist
      ? [artist]
      : [];

  return makeAlbumCandidate({
    artist: albumArtist,
    artists: albumArtists,
    album,
    albumArtist,
    albumArtists,
    year,
    musicbrainzAlbumId: request.musicbrainzAlbumId,
    musicbrainzArtistId: request.musicbrainzArtistId,
    discogsReleaseId: request.discogsReleaseId,
    discogsArtistId: request.discogsArtistId,
    tracks: request.tracks,
    source: "folder",
  });
}
